pkg/addon/templateagent/template_agent.go

```diff
@@ -233,3 +232,38 @@ func (a *CRDTemplateAgentAddon) getDesiredAddOnTemplateInner(
 
 	return template.DeepCopy(), nil
 }
+
+// TemplateAgentRegistrationNamespaceFunc reads deployment resource in the manifests and use that namespace as the default
+// registration namespace. If addonDeploymentConfig is set, uses the namespace in it.
+func (a *CRDTemplateAgentAddon) TemplateAgentRegistrationNamespaceFunc(addon *addonapiv1alpha1.ManagedClusterAddOn) (string, error) {
+	template, err := a.getDesiredAddOnTemplateInner(addon.Name, addon.Status.ConfigReferences)
+	if err != nil {
+		return "", err
+	}
+
+	// pick the namespace of the first deployment
+	var desiredNS = "open-cluster-management-agent-addon"
+	for _, manifest := range template.Spec.AgentSpec.Workload.Manifests {
+		object := &unstructured.Unstructured{}
```

Comment:
we have a helper func in addon framework to check if it is a deployment.
https://github.com/open-cluster-management-io/addon-framework/blob/main/pkg/utils/probe_helper.go#L97-L118